src/data_source/sql.rs

```diff
@@ -0,0 +1,987 @@
+/// migration is to be ordered relative to other migrations, and `name` is a descriptive name for
+/// the migration.
+///
+/// `path` should be an absolute path. It is possible to give a path relative to the root of the
+/// invoking crate by using environment variable expansions and the `CARGO_MANIFEST_DIR` environment
+/// variable.
+///
+/// As an example, this is the invocation used to load the default migrations from the
+/// `hotshot-query-service` crate. The migrations are located in a directory called `migrations` at
+/// the root of the crate.
+///
+/// ```
+/// # use hotshot_query_service::data_source::sql::{include_migrations, Migration};
+/// let migrations: Vec<Migration> = include_migrations!("$CARGO_MANIFEST_DIR/migrations").collect();
+/// assert_eq!(migrations[0].version(), 10);
+/// assert_eq!(migrations[0].name(), "init_schema");
+/// ```
+#[macro_export]
+macro_rules! include_migrations {
+    ($dir:tt) => {
+        $crate::data_source::sql::include_dir!($dir)
+            .files()
+            .map(|file| {
+                let path = file.path();
+                let name = path
+                    .file_name()
+                    .and_then(std::ffi::OsStr::to_str)
+                    .unwrap_or_else(|| {
+                        panic!(
+                            "migration file {} must have a non-empty UTF-8 name",
+                            path.display()
+                        )
+                    });
+                let sql = file
+                    .contents_utf8()
+                    .unwrap_or_else(|| panic!("migration file {name} must use UTF-8 encoding"));
+                Migration::unapplied(name, sql).expect("invalid migration")
+            })
+    };
+}
+
+/// The migrations requied to build the default schema for this version of [`SqlDataSource`].
+pub fn default_migrations() -> Vec<Migration> {
+    let mut migrations = include_migrations!("$CARGO_MANIFEST_DIR/migrations").collect::<Vec<_>>();
+    validate_migrations(&mut migrations, true).expect("default migrations are invalid");
+    migrations
+}
+
+/// Validate and preprocess a sequence of migrations.
+///
+/// * Ensure all migrations have distinct versions
+/// * Ensure migrations are sorted by increasing version
+/// * If validating [`default_migrations`] (i.e. `default` is `true`), additionally check that all
+///   migration versions are multiples of 10, so that custom migrations can be inserted in between.
+fn validate_migrations(migrations: &mut [Migration], default: bool) -> Result<(), Error> {
+    migrations.sort_by_key(|m| m.version());
+
+    // Check version uniqueness.
+    for (prev, next) in migrations.iter().zip(migrations.iter().skip(1)) {
+        if next <= prev {
+            return Err(Error::msg(format!(
+                "migration versions are not strictly increasing ({prev}->{next})"
+            )));
+        }
+    }
+
+    if default {
```

Comment:
I think this would be clearer with separate functions to check version spacing, check uniqueness, `validate_default_migrations` and `validate_custom_migrations` to remove the `default` boolean parameter that turns this function into one that does two things.